### PR TITLE
Fix references for `ResultCollection.extra` and improve `str` and `repr`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes an issue where the `ResultCollection.extra` return dict was not set back to the object in the default case.

Also improve its string representations.

Found while debugging #8613

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: tests/core/test_metric_result_integration.py [hunk 1]
```diff
@@ -132,13 +132,28 @@ def test_result_metric_integration():
 
         assert epoch_log == {"b": cumulative_sum, "a_epoch": cumulative_sum}
 
+    result.minimize = torch.tensor(1.0)
+    result.extra = {}
     assert str(result) == (
-        "ResultCollection(True, cpu, {"
+        "ResultCollection("
+        "minimize=1.0, "
```
Comment (tchaton) [inline tests/core/test_metric_result_integration.py:139]: Personal taste, but I prefered the previous one :)

**Review by tchaton — APPROVED** [review]

File: tests/core/test_metric_result_integration.py (hunk 1 above)
Comment (carmocca) [reply tests/core/test_metric_result_integration.py:139]: We still have the previous one as `__repr__`.

I found this splitting these made debugging easier. You don't want to show all the details in `__str__`
